File: Top_100_Liked_Questions/207._Course_Schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # dfs traversal in a dag
        g = {}
        for dst, src in prerequisites:
            if src in g:
                g[src].append(dst)
            else:
                g[src] = [dst]
        
        visited = set()
        visiting = set()
        
        def hasCycle(src):
            if src in visiting:
                return True
            
            if src in visited:
                return False
            
            visiting.add(src)
            
            for dst in g.get(src, []):
                if hasCycle(dst):
                    return True
            
            visiting.remove(src)
            visited.add(src)
            return False
        
        for src in g.keys():
            if hasCycle(src):
                return False
            
        return True

File: Top_100_Liked_Questions/207._Course_Schedule/solution.java
class Solution {
    public Set<Integer> visited;
    public Set<Integer> visiting;
    public Map<Integer, List<Integer>> map;

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        this.map = new HashMap<Integer, List<Integer>>();
        for (int[] prerequisite : prerequisites) {
            List<Integer> prereqs = map.getOrDefault(prerequisite[0], new ArrayList<Integer>());
            prereqs.add(prerequisite[1]);
            map.put(prerequisite[0], prereqs);
        }

        this.visited = new HashSet<Integer>();
        this.visiting = new HashSet<Integer>();
        while (this.map.size() > 0) {
            for (int course : this.map.keySet()) {
                if (!this.dfs(course))
                    return false;

                break;
            }
        }

        return true;
    }

    public boolean dfs(int course) {
        if (this.visited.contains(course))
            return true;

        if (this.visiting.contains(course))
            return false;

        if (this.map.containsKey(course)) {
            this.visiting.add(course);
            List<Integer> prereqs = this.map.get(course);
            for (int i=0; i<prereqs.size(); i++) {
                if (!this.dfs(prereqs.get(i)))
                    return false;
            }
            this.visiting.remove(course);
        }

        this.visited.add(course);
        this.map.remove(course);

        return true;
    }
}